### PR TITLE
Add Support for Per-Query-User-Information / Impersonation

### LINKED ISSUE #251: [Feature Request] Add capability to impersonate user (pass Grafana connected to Trino and execute request in his name)
Hello Everyone, 

**Some context**
As a user of Starburst, we need to be able to impersonate Grafana's connected users to execute request on Starburst.
Because we apply access policies with Ranger on Starburst and want to be able to apply the right access of the user who is connected to Grafana.
We don't want to use a generic service account, as we can't figure who exactly perform the request, and we need that information for two reasons:
- auditing
- applying the right access to the requested ressource at a user (Grafana's connected user) level

For now, the user configured on the Grafana's Data Source configuration is the one which perform the request on Starburst.

**So, what we are looking for:** 
- Configure the Data Source to activate impersonation of Grafana's connected user
- Impersonate all requests done by Grafana over Trino with the Grafana's connected user.

**Some search on what could be done (maybe) to achieve that**
Unfortunately, I don't have suffisiant knowlegde to make the dev myself for this. That's why I come here...

But here the things I think are needed to be done/implemented (even if I don't know exactly where):

- Add a checkbox like "impersonationNeeded" on the config page of the Grafana Trino DataSource (in which file, I don't know ?)  
- Then we have to get the connected user, maybe something like that in the Datasource.ts:  
```
impersonationNeeded: Boolean;
connectedUser: String;

constructor(....) {
	this.impersonationNeeded = instanceSettings.jsonData.impersonationNeeded || false;
	this.connectedUser = instanceSettings.username;
}
```
  
Then maybe a possibility to pass connected user to each Trino request::  
- And use those informations when submitting a request to Trino via HTTP connexion:
```
doRequest(options: any) {
    options.headers = options.headers || {};

    // Add the impersonation header if set
    if (this.impersonationNeeded) {
      options.headers['X-Trino-User'] = this.connectedUser;
    }

    return this.backendSrv.datasourceRequest[...];
  }
```

But does this header option be transferred to the Trino client when calling request, I don't know ?

Another possibility, is maybe adding this option in the Trino client in Go directly when executing a request:
https://github.com/trinodb/trino-go-client/blob/60351ff4fb2c3a2c1e67ea660ee37d286b2aeb15/README.md#system-access-control-and-per-query-user-information
```
db.Query("SELECT * FROM foobar WHERE id=?", 1, sql.Named("X-Trino-User", string(<connectedUser>)))
```

But I don't see how to pass <connectedUser> to the Trino GO Client instance....

I think this feature can useful for many Trino user.

Can someone know how to do that ?

Thanks,

Eric

### PR DESCRIPTION
**Pull request by FabianScheidt**
One of the awesome features in Trino is that the session user can be different from the authenticated user. This allows the authenticated user to impersonate other users.

This PR adds a settings that allows to impersonate the current Grafana user, so we can use the same connection and dashboards for users with different permissions (similar to what we can do with Apache Superset). If enabled, the current user login will be passed as [X-Trino-User](https://trino.io/docs/current/develop/client-protocol.html#client-request-headers).

Hope you find this useful and consider merging. I'm happy to make changes if needed.

Fixes #251

### CODE REVIEW

File: src/ConfigEditor.tsx [hunk 1]
```diff
@@ -10,13 +10,32 @@ interface State {}
...
           dataSourceConfig={options}
           onChange={onOptionsChange}
         />
+
+        <h3 className="page-heading">Trino</h3>
+        <div className="gf-form-group">
+          <div className="gf-form-inline">
+            <InlineField
+              label="Impersonate logged in user"
+              tooltip="If enabled, set X-Trino-User to the current Grafana user"
```
Comment (nineinchnick) [inline src/ConfigEditor.tsx:29]:
```suggestion
              tooltip="If enabled, set the Trino session user to the current Grafana user"
```

**Review by nineinchnick — APPROVED** [review]
This is great, I would have never figured it out myself, thanks!

Fixes #251

File: src/ConfigEditor.tsx (hunk 1 above)
Comment (FabianScheidt) [reply src/ConfigEditor.tsx]: Glad to hear that :)

I have applied your suggestion. Also I have signed the CLA, so I guess we'll have to wait for that to be processed.

**Comment by nineinchnick** [comment]
@cla-bot check

**Comment by FabianScheidt** [comment]
Seems like my CLA has been processed :)

@cla-bot check

**Comment by nineinchnick** [comment]
I'll test this manually, merge this and do a new release somewhere over the weekend.

**Comment by FabianScheidt** [comment]
Oh, if you need a test setup, here's mine: It's the [Example HTTP Connector](https://trino.io/docs/current/develop/example-http.html) with [File-based access control](https://trino.io/docs/current/security/file-system-access-control.html). If you build a dashboard for the _numbers_ table, _admin_ should be able to see all contents. If you impersonate _alice_, she should only see even numbers.

<details>
<summary>example-http.properties</summary>
<pre>
connector.name=example-http
metadata-uri=https://trino-http-example.s3.eu-central-1.amazonaws.com/example-metadata.json
</pre>
</details>

<details>
<summary>access-control.properties</summary>
<pre>
access-control.name=file
security.config-file=/etc/trino/rules.json
</pre>
</details>

<details>
<summary>rules.json</summary>
<pre>
{
  "tables": [
    {
      "user": "alice",
      "catalog": "example-http",
      "schema": "example",
      "table": "numbers",
      "privileges": ["SELECT"],
      "filter": "value % 2 = 0"
    },
    {
      "user": "alice",
      "privileges": ["SELECT"]
    },
    {
      "user": "admin",
      "privileges": ["SELECT"]
    }
  ],
  "impersonation": [
    {
      "original_role": "admin",
      "new_user": ".*",
      "allow": true
    }
  ]
}
</pre>
</details>

**Comment by nineinchnick** [comment]
1.0.7 was just accepted by Grafana and published in their catalog, thanks again!

**Pull request merged**